### PR TITLE
Section-2-fundamentals

### PR DESCRIPTION
**Pull request by stall84**
Close to finishing Variables section with notes. Added a guessing_game function with notes using the standard library.

### CODE REVIEW

**Comment by stall84** [comment]
ap

File: variables/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        let d = 50;
+        // In this block d will be 50
+    }
+    // In the outter scope d will be 10
+
+    let mut doggie = "woof!";
+    let doggie = "bark!";
+    // doggie will now be immutable 'bark!'
+
```
Comment (CleanCut) [inline variables/src/main.rs:36]: You never used the value of doggie from line 33 anywhere!

What I was demonstrating in the lecture was how to use shadowing to "change" the mutability of a variable (really, you're replacing the variable with another one, but there's no runtime cost so it's effectively changing it from our point of view).

```suggestion
    let mut doggie = "woof!";
    let doggie = doggie;
    // doggie will now be immutable 'woof!'

```

File: guessing_game/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    io::stdin() // We call the function stdin() from the (above) imported standard library. This function returns a terminal i/o instance object.
+        .read_line(&mut guess) // We call the read_line method on that returned terminal i/o instance. We pass the mutable empty string object (as a reference) to the readline i-o handler
+        // The ampersand (&) preceding our empty string object signifies we are passing this argument by reference.. because it's an object, we want to be able
+        // to access this object from anywhere without needing to copy that data into memory multiple times. Like variables, references are also immutable by default
+        // So we have to add (mut) the same as we would for variables.
+        // https://doc.rust-lang.org/book/ch02-00-guessing-game-tutorial.html#receiving-user-input:~:text=The%20%26%20indicates,references%20more%20thoroughly.)
+        .expect("Failed to read line..");
+    // read_line returns a Result enumeration.. A result enumeration can be in one of 2 states. OK and ERR ..
```
Comment (CleanCut) [inline guessing_game/src/main.rs:19]:
```suggestion
    // read_line returns a Result enumeration.. A result enumeration can be in one of 2 states. Ok and Err ..
```

File: guessing_game/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    // Read this excellent description. https://doc.rust-lang.org/book/ch02-00-guessing-game-tutorial.html#receiving-user-input:~:text=Values%20of%20the,the%20user%E2%80%99s%20input.
+    // Suffice to say the 'expect()' method allows you to write a statement to be
+    // displayed after gracefully crashing your program due to an unhandled or 'expected' (verbose in your string argument/description) error.
```
Comment (CleanCut) [inline guessing_game/src/main.rs:22]: Hehe, I'm not sure I'd describe crashing as _graceful_ per se. 😆

**Review by CleanCut — APPROVED** [review]
Nice! Looks like you're grasping the concepts. 😄

**Comment by stall84** [comment]
@CleanCut I need to replace the .git file! I apologize, I was wondering why this PR didn't show up on my github.
I'm loving the course though man!
 I really really really like how (coming from typescript.. some basic strongly typed elsewhere stuff) how Rust 'feels' with it's syntax .. it's very inviting coming from ES6, and Python..

**Pull request closed without merging**